### PR TITLE
Enabling find functionality on small devices

### PR DESCRIPTION
**Pull request by Thunderforge**
This is a work in progress fix to allow for find functionality when PDF.js is used on small devices, addressing #6191. No attempt was made to address the issues regarding foreign languages in #3683; this will likely be something that is done after find functionality is included on small devices.

At about 600px, find functionality is visible with some of the toolbar icons on the right side being moved to the overflow bar:

image

When the window shrinks to 570px, all elements with the `hiddenSmallView` CSS class are invisible. This no longer includes the Find functionality, but does include the next and previous buttons, per the discussion on the UX team where they pointed out that finger scrolling was far more likely to be used than a tiny button. Also, the find toolbar becomes extra tall. This will need more work, as there is a lot of empty space:

image

As the window gets smaller, the find toolbar checkboxes move to fill in the new space. Again, there is a lot of wasted space and improvements could be made:

image

When the window reaches 440px, the page label and the zoom dropdown menus are removed entirely:

image

I would appreciate feedback on these changes, and any suggestions to solve the issue with extra space on the find toolbar when it gets to double height.

### CODE REVIEW

File: web/viewer.html [hunk 1]
```diff
@@ -142,10 +142,12 @@
               <span data-l10n-id="find_next_label">Next</span>
             </button>
           </div>
-          <input type="checkbox" id="findHighlightAll" class="toolbarField">
```
Comment (Rob--W) [inline web/viewer.html]: Why did you introduce this new `<div>` and unused ID `findOptions`?

Comment (Thunderforge) [inline web/viewer.html]: The idea behind the new div was that it kept the checkboxes together as the screen size got smaller. It's not very useful to have a checkbox on one line and its label on another. The id was there for clarity. If there is a better solution< i'm happy to hear it.

Comment (Rob--W) [inline web/viewer.html]: Have you tried to put the input inside the label instead of wrapping the whole checkbox set in a div? If you put the input in the label, then the `for` attribute of the label is not needed any more. Verify that the the appearance is still OK after making this change though.

**Comment by timvandermeij** [comment]
This is going in the right direction, but I agree with @Rob--W that we should get rid of the added `div` element. Ideally we do not touch the HTML in this commit at all (except for changing classes) as it should not be necessary.

**Comment by Thunderforge** [comment]
To clarify about why I added the `<div>`, it was to avoid a problem like this:

image

I believe that Rob--W was suggesting I try something like this:

    <label class="toolbarLabel" data-l10n-id="find_highlight">
      <input type="checkbox" id="findHighlightAll" class="toolbarField" tabindex="94">
      Highlight all
    </label>
    <label class="toolbarLabel" data-l10n-id="find_match_case_label">
      <input type="checkbox" id="findMatchCase" class="toolbarField" tabindex="95">
      Match case
    </label>

However, this makes the checkboxes not appear correctly:

image

File: web/viewer.html [hunk 2]
```diff
@@ -142,10 +142,14 @@
               <span data-l10n-id="find_next_label">Next</span>
             </button>
           </div>
-          <input type="checkbox" id="findHighlightAll" class="toolbarField" tabindex="94">
-          <label for="findHighlightAll" class="toolbarLabel" data-l10n-id="find_highlight">Highlight all</label>
-          <input type="checkbox" id="findMatchCase" class="toolbarField" tabindex="95">
-          <label for="findMatchCase" class="toolbarLabel" data-l10n-id="find_match_case_label">Match case</label>
+          <label data-l10n-id="find_highlight">
```
Comment (Rob--W) [inline web/viewer.html]: The content of the element identified by `data-l10n-id="find_highlight"` will be replaced with the translation. So the checkbox will disappear with this change.

You should remove the attribute from here, wrap "Highlight all" in a `<span>` and put the attribute on the `<span>`. Same for the other checkboxs.

Comment (Thunderforge) [inline web/viewer.html]: Fixed. Good catch!

**Comment by timvandermeij** [comment]
@Snuffleupagus What is your opinion on the current version and code? Me, @Thunderforge and @Rob--W discussed this quite a bit on IRC (see the backlog if you want) and the following solution improves the responsiveness and usability quite a lot in my opinion. I'm curious to hear what you think, and if you have any improvements.

**Comment by arni2033** [comment]
Have you tested it with message "The end of document was reached, continued from the top" (I don't know exactly how that message look in English locale)
I also really think that those checkboxes, as well as span#findMsg message could be replaced with icons once the width is small enough

**Comment by Snuffleupagus** [comment]
> @Snuffleupagus What is your opinion on the current version and code?

Unfortunately changing a number of the `height: 32px;` instances to `height: auto;` seems to have resulted in various size/alignment issues.
For example: the `height` of the `findbar` now defaults to `33px` instead of `32px`, so that needs to be fixed. Another visual glitch that this PR introduces (again it's `height: auto;`) is that the initial height of the `loadingBar` changes, see the screen-shot below.
Do we actually, at least in this PR, need to change the `height` of anything *but* the `findbar`, since it causes various issues (and is going to require *very* careful manual testing to avoid breakage)?

loadingbar

---
Finally, we should make sure that this is landed at the beginning of a development cycle ([see dates here](https://wiki.mozilla.org/RapidRelease/Calendar#Future_branch_dates)), to avoid having to uplift any follow-up fixes.

File: web/viewer.html [hunk 3]
```diff
@@ -211,15 +217,15 @@
                   <span data-l10n-id="toggle_sidebar_label">Toggle Sidebar</span>
                 </button>
                 <div class="toolbarButtonSpacer"></div>
-                <button id="viewFind" class="toolbarButton group hiddenSmallView" title="Find in Document" tabindex="12" data-l10n-id="findbar">
+                <button id="viewFind" class="toolbarButton group" title="Find in Document" tabindex="12" data-l10n-id="findbar">
                    <span data-l10n-id="findbar_label">Find</span>
                 </button>
                 <div class="splitToolbarButton">
-                  <button class="toolbarButton pageUp" title="Previous Page" id="previous" tabindex="13" data-l10n-id="previous">
+                  <button class="toolbarButton pageUp hiddenSmallView" title="Previous Page" id="previous" tabindex="13" data-l10n-id="previous">
```
Comment (Snuffleupagus) [inline web/viewer.html]: Please add `hiddenSmallView` to the `<div class="splitToolbarButton">` just above instead, since adding it to each element seems unnecessary (and more prone to errors if you forget one child element).

File: web/viewer.css [hunk 4]
```diff
@@ -291,7 +295,7 @@ html[dir='rtl'] #toolbarSidebar {
 
 #toolbarContainer, .findbar, .secondaryToolbar {
   position: relative;
-  height: 32px;
+  height: auto;
```
Comment (Snuffleupagus) [inline web/viewer.css]: I think that you should revert this change, since it has negative consequences for e.g. the display of the `loadingBar` (and possibly other things as well, that's just what I've seen so far).

File: web/viewer.html (hunk 3 above)
Comment (Thunderforge) [inline web/viewer.html]: Fixed. I did need to use your additional CSS change to have it work.

**Comment by Thunderforge** [comment]
@arni2033 Here is what it looks like with the end of document message:

image

image

image

image

I'm not sure how others feel about this message moving around the way that it is.

Where were you suggesting to add `text-overflow: ellipsis;`. To `#findMsg`?

As for your other suggestion about icons for the checkboxes at small sizes, do you have a suggestion for what sort of icon would be used for "Highlight all" and "Match case"?

File: web/viewer.css (hunk 4 above)
Comment (Thunderforge) [inline web/viewer.css]: It is necessary for `.findbar` at least to have a height of `auto` so that it can grow when we need to rows of things. I suppose keeping that separate from #toolbarContainer and .secondaryToolbar would be something we can change.

Comment (Snuffleupagus) [inline web/viewer.css]: But you've already added that at https://github.com/mozilla/pdf.js/pull/6274/files#diff-5a36ed3216dce1e4da7e938ee4503ebdR379, which definitely should be enough!
This line, however, *needs* to be removed, since it breaks things (as explained above):
overlap

Comment (Thunderforge) [inline web/viewer.css]: Thanks for clarifying. I've committed and pushed the change. I'll squash it shortly.

**Comment by timvandermeij** [comment]
Looks better like this. Let's make a new preview.

/botio-linux preview

**Comment by timvandermeij** [comment]
I personally think adding ellipsis is not a great idea as that could make the message unclear. We also have enough space to put the full message. Icons are also a bit confusing in my opinion, especially since there is enough room for the labels.

Personally I think this PR is almost there except for the situation in https://cloud.githubusercontent.com/assets/6200170/8935639/5afc507e-3513-11e5-9dd8-616d0decad86.png. If anyone has ideas to prevent that from happening, please do tell :)

**Comment by Thunderforge** [comment]
@timvandermeij To be clear, what is the desired behavior that we want? Are we wanting the message to be left aligned below the find input field? Or something else?

**Comment by timvandermeij** [comment]
Yes, that does seem like the desired behavior to me.

**Comment by arni2033** [comment]
@Thunderforge 
Sorry I have no time to make normal answer OR read other posts, and also I'll have no internet connection for a week
I don't insist on any of my suggestions
2015 07 29 pdfjs

**Comment by Thunderforge** [comment]
So it doesn't look like anybody has been able to come up with a solution to that remaining problem in the last 4 weeks. Is there anything we can do to make sure that this work doesn't get lost?

**Comment by Rob--W** [comment]
@Thunderforge What exactly needs to be done before this PR can be merged?

**Comment by Thunderforge** [comment]
@Rob--W, there was the issue that @timvandermeij brought up saying that https://cloud.githubusercontent.com/assets/6200170/8935639/5afc507e-3513-11e5-9dd8-616d0decad86.png needed to be fixed. Instead it should be left aligned below the find input field, as in https://cloud.githubusercontent.com/assets/6200170/8935648/64659062-3513-11e5-8cc3-87ec6fc54150.png. Tim seemed to think that this was a dealbreaker, and unfortunately I haven't found a away around this, and nobody else has either as of yet.

**Comment by Rob--W** [comment]
@Thunderforge (see bottom for TL;DR)

The issue is caused by the difference in heights of the elements at the top row (which in turn is different across different browsers / operating systems because these heights depend on the heights of the button/checkbox elements, which is OS/browser-dependent).

In order to get floated elements to align to the left, the height + vertical margins of the last element(s) (=the one before the label) must be at least as big as the margin of the first element. Here is a snippet to quickly calculate the heights of the children of the findbar:

```js
[].map.call(document.getElementById('findbar').children, function(e) {
  var s = getComputedStyle(e);
  return e.offsetHeight + parseFloat(s.marginTop) + parseFloat(s.marginBottom);
});
```

On Firefox (Linux), the overall height of the checkbox container is 35px, which is bigger than the rest, so the next element starts at the start of the new line.
On Firefox (Windows 7), the overall height of the checkbox container is just 30, which is smaller than the 33px of the button container, so the next element is aligned to the right of the button container.

Firefox 40 (Linux):
firefox40-linux

Firefox 40 (Windows 7):
firefox40-win

**TL;DR** Change the styles such that the heights of the children of `#findbar` are all equal (margin top + margin bottom + height (+border and padding if box-sizing is content-box instead of border-box. we do not specify box-sizing, so the border and padding must be included in the calculation)).

**Comment by Thunderforge** [comment]
I've adjusted the margins for `splitToolbarButton`, which solves the issue without (as far as I can tell) introducing any side effects:

image

image

image

image

**Comment by Snuffleupagus** [comment]
> I've adjusted the margins for splitToolbarButton, which solves the issue without (as far as I can tell) introducing any side effects:

Unfortunately this change does seem to cause all `splitToolbarButton`s to be moved upwards, please see the following comparison (taken on Windows):

height_diff

File: web/viewer.css [hunk 5]
```diff
@@ -1084,7 +1093,6 @@ html[dir='rtl'] .verticalToolbarSeparator {
 }
 
 .toolbarLabel {
-  min-width: 16px;
```
Comment (Snuffleupagus) [inline web/viewer.css]: Is it actually necessary to remove this line?
If I understand correctly, you changed this in order to avoid adding a bunch of empty space at the right side of the `findbar` when `findMsg` isn't set. If that's correct, wouldn't adding the following rule achieve that in a better/simpler way?
```css
#findMsg:empty {
  display: none;
}
```

Comment (Thunderforge) [inline web/viewer.css]: This does appear to be a better solution, and it reduces even more extraneous space than the former solution. I've pushed a commit with this change.

**Comment by Thunderforge** [comment]
I've backed out of the commit that changed the height of all `splitToolbarButton`. I tried adding the following rule specific to the findbar's `splitToolbarButton`, but it didn't produce the desired effect:

    #findbar > .splitToolbarButton {
      margin-top: 2px;
      margin-bottom: 1px;
    }

image

**Comment by Snuffleupagus** [comment]
Even if this is now looking much better overall, there still appears to be an issue with the `findbar` becoming way too wide, using all the available window width, in some cases (Firefox 42 on Windows):

findbar-full-width

**Comment by Thunderforge** [comment]
@Snuffleupagus The issue you describe was brought up in IRC and @timvandermeij had said that this was acceptable. I would certainly be willing to make a change to prevent that, if we could find a way around it.

**Comment by Thunderforge** [comment]
Has there been any further analysis on this pull request? Is it ready to be committed or do further changes need to be made?

**Comment by timvandermeij** [comment]
@Snuffleupagus @Thunderforge Personally I'm fine with the patch as-is, but if we can find a way to solve the findbar extension that would be great. I have looked a bit into that before and I couldn't really find a suitable solution myself, but maybe @Snuffleupagus or @Rob--W have an idea to solve this?

**Comment by Rob--W** [comment]
@Thunderforge @timvandermeij 

Why is `vertical-align:bottom` used instead of `middle`? The difference in vertical alignment results in a weird label:
phrase-not-found-vertical-align-bottom

The checkboxes + result phrase moves to the next line because they're in a single box, and together they don't fit at the right of the find bar's top area. Try moving the status message outside the common container.

```
Before:
#findbar
  #findControls
    label.toolbarLabel
    label.toolbarLabel
    #findMsg

After:
#findbar
  #findControls
    label.toolbarLabel
    label.toolbarLabel
  #findMsg
```

**Comment by Thunderforge** [comment]
@Rob--W I finally got a chance to address your issues. I presume you were referring to `vertical-align:bottom` being in `#findMsg`. I've changed it to `vertical-align:middle`

As for moving the `#findMsg` outside of the div, this was done to avoid [an issue with alignment](https://cloud.githubusercontent.com/assets/6200170/8935639/5afc507e-3513-11e5-9dd8-616d0decad86.png) described earlier in this thread by @timvandermeij. Moving it out brings back that behavior. We were unable to find a solution to this aside from putting it into a `<div>` (I believe this was decided in IRC).

Ideally, we would like for it to appear [like this](https://cloud.githubusercontent.com/assets/6200170/8935648/64659062-3513-11e5-8cc3-87ec6fc54150.png) at all window sizes where wrapping must occur. However, we could not find a way to do that without either causing [an issue with alignment](https://cloud.githubusercontent.com/assets/6200170/8935639/5afc507e-3513-11e5-9dd8-616d0decad86.png) or an issue with [extra space](https://cloud.githubusercontent.com/assets/2692120/9801579/4846d354-5814-11e5-9dcb-6792f410a7b3.png). Unless there is an obvious solution that we are missing, I think we need to make a decision for which issue we are willing to live with.

**Comment by Snuffleupagus** [comment]
@Thunderforge I've not really had time to look into the case of the extraneous width of the Findbar when wrapping occurs (and I'm also starting to think that we *may* have to live with that).

However, it seems to me that there may be an easier way to avoid the vertical alignment issues that have been described in a number of the comments; [please see the following diff for an idea](https://github.com/mozilla/pdf.js/commit/b1b190188e450f3ee2671b474fa6d4e64672846e?w=1).
Since the height of certain elements in the Findbar can depend on the browser/OS, I'm not certain that just changing `margin-top` of some elements is guaranteed to always work.

**Comment by mazyvan** [comment]
Thanks men. But why this is still not officially on PDF.js ? 
Anyway this help me a lot

**Comment by timvandermeij** [comment]
Closing since the patch above provides an updated version that should resolve all issues mentioned in this thread. Nevertheless, thank you for the contribution!
